File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    public boolean pacific[][];
    public boolean atlantic[][];
    public boolean visi[][];
    
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> l = new ArrayList<>();
        int m = heights.length;
        int n = heights[0].length;
        pacific = new boolean[m][n];
        atlantic = new boolean[m][n];
        visi = new boolean[m][n];
        for (int i = 0;i<m;i++)
        {
            findSea(heights, m, n, i, 0, heights[i][0], pacific);
            findSea(heights, m, n, i, n-1, heights[i][n-1], atlantic);
        }
        for (int j = 0;j<n;j++)
        {
            findSea(heights, m, n, 0, j, heights[0][j], pacific);
            findSea(heights, m, n, m-1, j, heights[m-1][j], atlantic);
        }
        for (int i = 0;i<m;i++)
        {
            for (int j = 0;j<n;j++)
            {
                if (pacific[i][j] == true && atlantic[i][j] == true)
                {
                    List<Integer> ll = new ArrayList<>();
                    ll.add(i);
                    ll.add(j);
                    l.add(new ArrayList<>(ll));
                }
            }
        }
        return l;
    }
    
    public void findSea(int[][] heights, int m, int n, int i, int j, int h, boolean[][] visited)
    {
        if (i < 0 || j < 0 || i == m || j == n || heights[i][j] < h)
        {
            return;
        }
        if (visi[i][j] == true)
        {
            return;
        }
        if (visited[i][j] == true)
        {
            return;
        }
        visited[i][j] = true;
        visi[i][j] = true;
        findSea(heights, m, n, i, j+1, heights[i][j], visited);
        findSea(heights, m, n, i+1, j, heights[i][j], visited);
        findSea(heights, m, n, i, j-1, heights[i][j], visited);
        findSea(heights, m, n, i-1, j, heights[i][j], visited);
        visi[i][j] = false;
        return;
    }
    
    
}

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        res = []
        rows = len(heights)
        cols = len(heights[0])
        tempp = [[False for _ in range(cols)] for _ in range(rows)]
        tempa = [[False for _ in range(cols)] for _ in range(rows)]
        visitp = [[False for _ in range(cols)] for _ in range(rows)]
        visita = [[False for _ in range(cols)] for _ in range(rows)]
        for i in range(rows):
            tempp[i][0] = True
            tempa[i][cols - 1] = True
        for j in range(cols):
            tempp[0][j] = True
            tempa[rows - 1][j] = True

        def dpp(i, j, height):
            nonlocal visitp
            if i < 0 or i >= rows or j < 0 or j >= cols or visitp[i][j] == True or heights[i][j] < height:
                return
            tempp[i][j] = True
            visitp[i][j] = True
            # print(i, j, height)
            dpp(i, j + 1, heights[i][j])
            dpp(i + 1, j, heights[i][j])
            dpp(i, j - 1, heights[i][j])
            dpp(i - 1, j, heights[i][j])
            return
        
        def dpa(i, j, height):
            nonlocal visita
            if i < 0 or i >= rows or j < 0 or j >= cols or visita[i][j] == True or heights[i][j] < height:
                return
            tempa[i][j] = True
            visita[i][j] = True
            dpa(i, j + 1, heights[i][j])
            dpa(i + 1, j, heights[i][j])
            dpa(i, j - 1, heights[i][j])
            dpa(i - 1, j, heights[i][j])
            return
        
        for i in range(rows):
            for j in range(cols):
                # print(i, j, tempp[i][j])
                if tempp[i][j] == True:
                    dpp(i, j, heights[i][j])
                if tempa[i][j] == True:
                    dpa(i, j, heights[i][j])
        
        for i in range(rows):
            for j in range(cols):
                if tempp[i][j] and tempa[i][j]:
                    res.append([i, j])

        return res